### PR TITLE
Add s3 storage backend

### PR DESCRIPTION
**Pull request by rbarry82**
## Issue
Add the ability to use s3 storage.

## Solution
Use the `object_storage` library from s3proxy for now

Shuffle methods in `src/charm.py` so they're ordered next to their callers, where possible. Rename a couple to have them reflect what they actually do

### CODE REVIEW

File: src/charm.py
Comment (sed-i) [inline src/charm.py:227]: Iiuc, this could be error-prone due to https://github.com/canonical/operator/issues/888

File: src/charm.py
Comment (sed-i) [inline src/charm.py:229]: What will happen to mimir/data if after `s3 broken` it is restarted with a different config (filesystem instead of s3)? Or vice versa?
I.e. do we need to manually migrate anything?

File: tests/integration/test_object_storage.py
Comment (sed-i) [inline tests/integration/test_object_storage.py:57]: Should probably do something with `block_until("/prometheus/labels" not empty)`

File: tests/integration/test_object_storage.py
Comment (sed-i) [inline tests/integration/test_object_storage.py:59]: Should we move this assertion to the previous function?

File: tests/integration/test_object_storage.py
Comment (sed-i) [inline tests/integration/test_object_storage.py:39]: How come you don't need to config it?
Can we have an indirect way to confirm our config in results in mimir attempting to user object storage?

File: lib/charms/s3proxy_k8s/v0/object_storage.py
Comment (rbarry82) [reply lib/charms/s3proxy_k8s/v0/object_storage.py:2]: The Apache license does not require this. A single copy at the top level of the repo is sufficient, legally, and anything else is a nicety. No linter is needed. 

The referenced library also has the license at the top level of the repo.

File: metadata.yaml
Comment (rbarry82) [reply metadata.yaml:49]: No. We are attempting to keep compatibility with the data platform lib/interfaces

File: pyproject.toml
Comment (rbarry82) [reply pyproject.toml:41]: No. It uses pyright anyway, and this issue will disappear when all of our charms are migrated. Using `follow_imports = "silent"` matches all of our other repos, too.

File: src/charm.py
Comment (rbarry82) [reply src/charm.py:102]: Is it "behind the scenes"? If they relate to it? The idea that software will use object storage if configured is normal, not unusual

File: src/charm.py
Comment (rbarry82) [reply src/charm.py:229]: The data is lost. "don't do that". Mimir has tools to dump and load, but they're not covered here. I also don't expect them to be. 

Note that loki, tempo, and phlare will have identical scenarios and problems. It's just a new thing for us.

File: tests/integration/test_object_storage.py
Comment (rbarry82) [reply tests/integration/test_object_storage.py:39]: It's preconfigured. If no key/secret is set, it generates one. Is there a way other than the actual running Mimir config (via endpoint) that you feel would be a better way of confirming it?

File: tests/integration/test_object_storage.py
Comment (rbarry82) [reply tests/integration/test_object_storage.py:57]: There are other labels also. We know the default push timer of avalanche, and we could configure it anyway. It's 60 seconds. 75 gives a small buffer. What will `block_until` buy us?

File: lib/charms/s3proxy_k8s/v0/object_storage.py
Comment (facundobatista) [reply lib/charms/s3proxy_k8s/v0/object_storage.py:2]: I also think a linter is not a good fit here, legal disclaimer headers may vary wildly and people would have good reasons to have it their ways.

File: src/charm.py
Comment (dstathis) [reply src/charm.py:229]: Also note the data presumably isn't lost. It is just no longer accessible to mimir.

**Review by dstathis — APPROVED** [review]

File: src/charm.py
Comment (PietroPasotti) [inline src/charm.py:306]: doesn't look like it's ever returning anything.

File: src/charm.py
Comment (PietroPasotti) [inline src/charm.py:314]: I've made it a habit to always `raise X from e` whenever I'm wrapping errors, so as not to lose any trace info. Maybe you're catching this event already someplace else, but I think it'd be good practice anyway.

File: src/charm.py
Comment (PietroPasotti) [inline src/charm.py:337]: rules_file is `str` then? I'd give the variable a different name then, and update the docstring accordingly. A type hint would also be nice.

File: src/charm.py
Comment (PietroPasotti) [inline src/charm.py:339]: you're not scared of deploying the tactical walrus, yet you're avoiding to use f-strings? :P

File: src/charm.py
Comment (PietroPasotti) [inline src/charm.py:233]: any chance that `s3["endpoint"]` is badly formatted? Should we try/catch or do some more guarded casting?

**Review by PietroPasotti — APPROVED** [review]
Looks good!
Some minor comments, doesn't need a second review from me.

File: lib/charms/s3proxy_k8s/v0/object_storage.py
Comment (sed-i) [reply lib/charms/s3proxy_k8s/v0/object_storage.py:2]: - This charm lib is taken from another charm, without attribution.
- This charm lib could have a different license from the charm that vendored it.

Here we know what's what and it's all Apache-2, but that is besides the point: we have an ecosystem problem.

File: src/charm.py
Comment (sed-i) [reply src/charm.py:102]: Can't say. Is storage a big enough of a deal?
Currently in metadata we have a top level section dedicated to it.
Could be friendly to mention it there if one expects to find all storages listed under storage.
:shrug:

File: src/charm.py
Comment (sed-i) [reply src/charm.py:175]: This code isn't new, just shuffled so appears green.
I added it in a previous commit.
"A balance of speculative future-proofing and readability".

File: src/charm.py
Comment (sed-i) [reply src/charm.py:229]: I think this should be documented (README?) as it is very easy to lose (access to) data this way unintentionally.

File: src/charm.py
Comment (sed-i) [reply src/charm.py:339]: `%s` in logger funcs is [more efficient](https://docs.python.org/3/howto/logging.html#optimization).

File: tests/integration/test_object_storage.py
Comment (sed-i) [reply tests/integration/test_object_storage.py:39]: Sure, if you can confirm from an endpoint that mimir is doing s3 now then great.

File: metadata.yaml
Comment (simskij) [reply metadata.yaml:49]: I can give my reason for agreeing with "no": The storage suffix does not add any value as s3 is an established term that most who will want to use it will already understand.

File: src/charm.py
Comment (simskij) [reply src/charm.py:229]: I strongly agree with Leon that this needs to be documented. Preferably we would also log whenever we detect potential data abandonment.

File: lib/charms/s3proxy_k8s/v0/object_storage.py
Comment (rbarry82) [reply lib/charms/s3proxy_k8s/v0/object_storage.py:2]: The Apache license doesn't require this. The file is not modified. This is not a legal problem, or an ecosystem problem.

The expectation is that _developers_ will follow the guidelines for appropriately licensed inclusions (`NOTICE` at the top level or other) where needed. There's not a meaningful way for CI or a linter to discern which files may or may not be pulled from elsewhere, what their parent repo is, what license it maybe under, and what that license requires.

File: src/charm.py
Comment (rbarry82) [reply src/charm.py:102]: `metadata.yaml` is for Juju to construct an object for the appropriate provider, and is not orthogonal to object storage anyway. The PV will still be allocated and actually used (for local evaluation of alerting/recording rules, local TSDB compaction, etc).

File: src/charm.py
Comment (rbarry82) [reply src/charm.py:233]: Not once CRI is in place. Very few charms implement this interface right now, and none are broken, but `http://` is part of the (current) definition.

File: src/charm.py
Comment (rbarry82) [reply src/charm.py:337]: Agreed. This is just a re-shuffle. I'll follow back through for cleanup.

File: src/charm.py
Comment (rbarry82) [reply src/charm.py:314]: Agreed. This is just a re-shuffle. I'll follow back through for cleanup.

**Pull request merged**
